### ISSUE TITLE
 security issue: gluon.tools.Expose will follow symlinks

### ISSUE THREAD

**Issue by chenl** [web2py/web2py#1261]
Consider `controller/default.py` with:

```
def expose_tmp_folder():
    return dict(files=Expose('/tmp'))
```

If a user can gain write access to `/tmp` and can perform `ln -s /etc /tmp/link-to-etc`, expose_tmp_folder will happily follow that link and expose `/etc` as well.

I suggest to examine the symlink and allow it only if the target is also under base.

I attach my solution as a patch. I don't do this as a pull request because I haven't tested it sufficiently.

[web2py-gluon-tools-expose-security.patch.txt](https://github.com/web2py/web2py/files/201318/web2py-gluon-tools-expose-security.patch.txt)

**Comment by leonelcamara** [comment]
Humm I don't know, the user should know better, the user also shouldn't give permission to whatever is running web2py to access files it should not, nor should it expose directories with symlinks to places he doesn't want to expose. This is a bit like saying that Expose is a security risk in itself because a dumb user could expose /etc directly.

**Comment by chenl** [comment]
I agree that if all the other security measures are perfectly implemented this typically will not be a problem, however when considering security one should implement multiple layers if possible.

The gluon.tools.Expose can be thought of as a chroot environment, and in such a case allowing symlinks to point out of it is breach of the chroot.

Also I do have a scenario where this becomes a problem. I use a web2py application as a management console of the localhost. I also have a user with very limited shell access on that host. I would like to Expose a public folder, but if symlinks are allowed that limited user can access the rest of the system by creating symlinks in that public folder.

I realise that this is an obscure use case, but it may demonstrate the complexity of security issues, and I think that the limitation of disallowing symlinks that point outside the base directory is a reasonable limitation, that keep the intent of the Expose class.

Please note that the Expose class already enforces some arbitrary policy, e.g. hiding files that have the string "private" somewhere in their name (but the user might be able to show such "private" files by manipulating the URL line).

I believe that my solution (that now have unit-tests, and that I will soon send as a pull request) is less arbitrary and more robust.

**Comment by BuhtigithuB** [comment]
Hello Chen,

Can it be optional with a warning and default to prevent the risk you are
explaining?

**Comment by chenl** [comment]
@BuhtigithuB, it is not hard to make it optional, and I agree it should be more secure by default. How do you suggest to do the warning? Is the doc string sufficient?

**Comment by BuhtigithuB** [comment]
At startup... I think a simple print statement that would run once could do
it...

Here how other stuff are printed out at start up :

https://github.com/web2py/web2py/blob/a0bbd7885a34f7974102e065c1a0639043cb62bb/gluon/widget.py#L1111

**Comment by chenl** [comment]
@BuhtigithuB, I am confused.

Are you suggesting that when the server is started, there should be some sort of a scan that find if any application calls `gluon.tools.Expose(..., allow_symlinks_out=True)` in any function of any controller?

**Comment by BuhtigithuB** [comment]
Not at startup as like you suggest it would implies some sort of scan... I
just point to this so you can see how stuff get printed out...

I am not sure how this should work, just make suggestion... Also, maybe
just a report in web2py log...

**Comment by BuhtigithuB** [comment]
Info about logging in web2py : http://web2py.com/books/default/chapter/29/04/the-core#Logging

**Comment by chenl** [comment]
Just send a pull request [fix issue#1261: security issue: gluon.tools.Expose will follow symlinks](https://github.com/web2py/web2py/pull/1277)

**Comment by chenl** [comment]
with the merge of pr#1357 this issue should be closed.
